File: HF1/khf1.erl

-module(khf1).
-author('czipobence@gmail.com').
-vsn('2016-09-22').
-export([feldarabolasa/2]).
%-compile(export_all).


%% @type matrix() = [row()].
%% @type row() = [any()].
%% @type parameter() = {subRows(), subCols()}.
%% @type subRows() = integer().
%% @type subCols() = integer().
%% @spec khf1:feldarabolasa(Mx::matrix(), P::parameter()) -> LL::[[any()]].
%%   Az LL lista az Mx mátrix P paraméterû feldarabolása.
feldarabolasa(Mx,{W,H}) -> [flatten(L) || L <- split(Mx, W, H)].

%% @spec khf1:split(Mx::matrix(), W::subRows(), H::subCols()) -> LM::[matrix()]
%% Az LM lista az Mx mátrix {W,H} paraméterû darabolásakor keletkezõ
%% mátrixok listája
split(Mx, W, H) -> split(Mx, W,H, countW(Mx,W), countH(Mx,H)).

%% @spec khf1:split(Mx::matrix(), W::subRows(), H::subCols(), CW::integer(), CH::integer()) -> LM::[matrix()]
%% Az LM lista az Mx mátrix {W,H} paraméterû darabolásakor keletkezõ
%% mátrixok listája. A mátrixot oszlopai mentén CW, sorok mentén CH
%% darabra vágja, így a lista elemszáma CW * CH lesz.
split(Mx,W, H,CW,CH) -> 
	[submatrix(Mx,X*W+1,W,Y*H+1,H) ||
											X <- lists:seq(0, CW-1),
											Y <- lists:seq(0, CH-1)]. 

%% @spec khf1:submatrix(Mx::matrix(), StartX::integer(), Width::subRows(), StartY::integer(), Height::subCols()) -> M::matrix()
%% M mátrix az Mx mátrix olyan sor és oszlopfolytonos részmátrixa,
%% ami a StartX oszloptól Width oszlop szélesen,
%% és a StartY sortól Height sor hosszan tart 									
submatrix(Mx, StartX, Width, StartY, Height) ->
	[sublist(L, StartY, Height) || L <- sublist(Mx, StartX, Width)].

%% @spec khf1:sublist (L0::[term()], S::integer, N::integer) -> L::[term()]
%% Az L lista az L0 lista N hosszúságú részlistája az S. elemmel kezdve
sublist(_, 0, _) -> [];
sublist(_, _, 0) -> [];
sublist([], _, _) -> [];
sublist(L0, S, N) -> take( drop(L0, S-1),N). 

%% @spec khf1:take(L0::[term()], N::integer()) -> L::[term()]
%% Az L lista az L0 lista N hosszú prefixuma
take([], _) -> [];
take(_, 0) -> [];
take([H|T], N) -> [H | take(T, N-1)].

%% @spec khf1:drop(L0::[term()], N::integer()) -> L::[term()]
%% Az L0 lista olyan szuffixuma L, amely az L0 elsõ N elemét
%% nem tartalmazza
drop([],_) -> [];
drop(L0,0) -> L0;
drop([_|T], N) -> drop(T, N-1).  

%% @spec khf1:flatten(Mx::matrix()) -> LL::[any()].
%% Az LL lista az Mx mátrix kilapítva
flatten(Mx) -> flatten(Mx, []).


%% @spec khf1:flatten(Mx::matrix(), L::[term()]) -> L1::[term()].
%% Az L1 lista az Mx mátrix kilapítva és L lista mögéfûzve
flatten([],L) -> L;
flatten([H|T], L) when is_list(H) -> flatten(H, flatten(T,L));
flatten([H|T], L) -> [H | flatten(T,L)].

%% @spec khf1:countW(L::[term()], W::integer()) -> CW::integer()
%% L listát W hosszú darabokra vágva CW darab lista keletkezik
countW(L, W) -> countW(L, W, 0, 0).


%% @spec khf1:countW(Mx::matrix(), H::integer()) -> CH::integer()
%% Mx mátrix oszlopait H hosszú darabra vágva CH darab oszlop keletkezik
countH([Hd|_], H) when is_list(Hd) -> countW(Hd, H).

%% @spec khf1:countW(L::[term()], W::integer(), Rem::integer(), Acc::integer()) -> CW::integer()
%% CW Acc hozzáadva ahhoz, hogy L lista Rem elemét elhagyva hány V
% hosszú darabra vághtató
countW([], _, _, Acc) -> Acc;
countW(L, W, 0, Acc) -> countW(L, W, W, Acc+1); 
countW([_|T], W, Rem, Acc) -> countW(T, W, Rem-1, Acc).

File: KHF/HF1/khf1.pl
% :- type matrix == list(row).
% :- type row == list(any).
% :- type parameter ---> subRows-subCols.
% :- type subRows == int.
% :- type subCols == int.
% :- pred feldarabolasa(+matrix, +parameter, ?list(list(any))).
% feldarabolasa(Mx, P, LL): Az LL lista az Mx mátrix P paraméterű feldarabolása.
feldarabolasa(Mx, R-C, LL) :- 
	feldarabolasa(Mx,R,C,LLt),
	flattenNextLvl(LLt,LL).


% feldarabolasa(Mx, R,C, LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása
feldarabolasa(Mx, R, C, LL):- 
	countR(Mx, R, Cr, 0),
	countC(Mx, C, Cc, 0),
	split(Mx, R, C, Cr, Cc, LL).


% split(Mx, R,C, Cr, Cc, LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása,
% sorok mentén legfeljebb Cr, oszolopok mentén legfeljebb Cc részre vágva
split([Hmx|Tmx], R,  C, Cr, Cc, LL) :-
	split([Hmx|Tmx], R, C, Cr, Cc, 0, 0, LL).

% split(Mx, R,C, Cr, Cc, Cntr, Cntc LL): Az LL lista az Mx mátrix R-C paraméterű mátrixokra darabolása,
% sorok mentén legfeljebb Cr, oszolopok mentén legfeljebb Cc részre vágva
% a sorok szerint Cntr, az oszlopok szerint Cntc-dik részmátrixtól kezdve
split(_, _, _, Cr, _, Cr, 0, []).
split([H|T],R,C,Cr,Cc,Cntr,Cc,L) :-
	Cntr1 is Cntr+1,
	split([H|T],R,C,Cr,Cc,Cntr1,0,L).
split([Hmx|Tmx], R,  C, Cr, Cc, Cntr, Cntc, [H1|T1]) :- 
	Cntr < Cr,
	Cntc < Cc,
	Cntc1 is Cntc + 1,
	StartC is Cntc * C + 1,
	StartR is Cntr * R + 1,
	submatrix([Hmx|Tmx], StartC, C, StartR, R, H1),
	split([Hmx|Tmx], R, C, Cr, Cc, Cntr, Cntc1, T1).

% submatrix(Mx, StartC, Width, StartR, Height, Smx):
%% Smx mátrix az Mx mátrix olyan sor és oszlopfolytonos részmátrixa,
%% ami a StartC oszloptól Width oszlop szélesen,
%% és a StartR sortól Height sor hosszan tart 
submatrix(Mx, StartC, Width, StartR, Height, Smx) :-
	sublist(Mx, StartR, Height, Lt),
	cutAll(Lt, StartC, Width, Smx).

% cutAll(Mx, StartR, Height, M) M mátrix Mx olyan részmátrixa, mely annak
% Height sorát tartalmazza sorfolytonosan a StartR. sortól kezdve	
cutAll([], _, _, []).
cutAll([H|T], StartR, Height, [HSmx|TSmx]) :-
	isList(H),
	sublist(H,StartR,Height, HSmx),
	cutAll(T, StartR, Height, TSmx).

% sublist(L, S, N, L1): Az L1 lista az L lista N hosszúságú részlistája az S. elemmel kezdve
sublist(L, S, N, L1) :-
	S1 is S -1,
	drop(L,S1,Ltemp),
	take(Ltemp,N,L1).

% take(L0, N, L): Az L lista az L0 lista N hosszú prefixuma
take([], _, []).
take([_|_], 0, []).
take([H|T], N, [H | Taken]) :-
	N > 0,
	N1 is N-1,
	take(T, N1, Taken).

% drop(L0, N, L): Az L0 lista olyan szuffixuma L, amely az L0 elsõ N elemét
%% nem tartalmazza
drop([], _, []).
drop([H|T], 0, [H|T]).
drop([_|T], N, Dropped) :-
	N > 0,
	N1 is N-1,
	drop(T,N1,Dropped).

% countC(Mx, C, Cc, Acc): Cc = Acc + ahány részre esik Mx mátrix C oszloponként vágva
countC([H|_], C, Cc, 0) :- 
	isList(H),
    countR(H,C,Cc,0).

%countR(Mx, R, Cr, Acc): Cr = Acc + ahány részre esik Mx mátrix R soronként vágva
countR([], _, 0, _).
countR([H|T], R, Cr, 0) :- 
	countR([H|T],R,UjCr,R),
	Cr is UjCr + 1.
countR([_|T], R, Cr, Rem) :-
	Rem > 0,
	UjRem is Rem -1,
	countR(T, R, Cr, UjRem).

% flattenNextLvl(L,Fl): Fl az L lista elemeit taralmazza, mindet kilapítva
flattenNextLvl([],[]).
flattenNextLvl([H0|T0],[H1|T1]):-
	flatten(H0,H1),
	flattenNextLvl(T0,T1).
	

% flatten(L,Fl) : Fl L lista kilapítva
flatten(L, Fl):-
  flatten(L, Fl, []).

% flatten(L, Fl, Acc) Fl az L lista kilapítva és Acc elé fűzve
flatten([], Fl, Fl).
flatten(H, [H|Fl], Fl):-
  \+ isList(H).
flatten([H|T], Fl, Acc):-
  flatten(H, Fl, L1),
  flatten(T, L1, Acc).

% isList(L) L egy tetszőleges lista (annak eldöntésére, hogy L lista-e)
isList([]).
isList([_|_]).
